### ISSUE TITLE
[GJEP] Add combinators for changing choicemap address scheme

### ISSUE THREAD

**Issue by georgematheos** [genjax-community/genjax#962]
Possible feature for v0.3.0.

Add a set of combinators which help change the address structure of a generative function.

Possible combinators:

1. `ChangeAddress(gen_fn, addr_transform, sub_addr=None)`. This runs `gen_fn`, but produces choicemaps where address `addr` is changed to `addr_transform(addr)`. If sub_addr is provided, this changes the addresses nested below sub_addr. \[It's possible we need another argument, which is the inverse to `addr_transform`, to implement this.\]
2. `ElideAddress(gen_fn, addr)`. This moves moves any sub-choicemaps at address `addr` to the top level.

There may be other useful combinators, and there are design questions to consider with these that I have proposed here.

Example use case:

```{python}
@genjax.static_gen_fn
def sequence(T):
  init = init_model() @ "init"
  step = genjax.unfold_combinator(step_model) @ "step"
 
nicely_addressed = ChangeAddress(sequence, lambda a: 0 if a == "init" else a)
nicely_addressed = ChangeAddress(nicely_addressed, lambda t : t + 1, "step") # move addrs ("step", 0) -> ("step", 1) ; ("step", 1) -> ("step", 2)
nicely_addressed = ElideAddress(nicely_addressed, "step") # move addrs below "step" to the top level
 
tr = nicely_addressed.simulate(key, (5,))
console.print(tr.get_choices())
# Printout should have addresses (0, sub_addresses_of_init...), (1, sub_addresses_of_step...), (2, sub_addresses_of_step...)
```

**Comment by femtomc** [comment]
Will consider this as part of discussion surrounding changes to the GFI.

**Comment by sritchie** [comment]
Closed, superceded by 0.4.0 / current nightly.
